refactor(patient registry)

lomamech opened the pull request:
refactor(patient registry)  …
- Ingrate grid save state
- Add dropdownfor for Column configuration, Downlowas view as CSV PDF
- Remove bhRenderer Dropdown

Update the standard for reduce technical debt and give required features
- Use the latest FilterService
- Update test E2E

closes #1748

lomamech committed:
refactor(patient registry)
- Ingrate grid save state
- Add dropdownfor for Column configuration, Downlowas view as CSV PDF
- Remove bhRenderer Dropdown

lomamech committed:
refactor(Patient registry)
Update the standard for reduce technical debt and give required features
- Use the latest FilterService
- Update test E2E

lomamech committed:
refactor(Label variable test E2E)

jniles commented on client/src/modules/account_statement/account_statement.html:
```diff
@@ -22,12 +22,12 @@
             <li role="menuitem">
               <a href data-method="print" id="print" ng-click="AccountStateCtrl.exportPdf()">
-                <i class="fa fa-file-pdf-o"></i> <span translate>FORM.BUTTONS.EXPORT</span> PDF
+                <i class="fa fa-file-pdf-o"></i> <span translate>FORM.BUTTONS.PDF_EXPORT</span>
```
Nice improvement :+1:

jniles commented on client/src/modules/patients/patients.service.js:
```diff
@@ -22,23 +22,22 @@ PatientService.$inject = [
+  Documents, Visits, Filters, AppCache, Periods, Api, $httpParamSerializer) {
   var baseUrl = '/patients/';
-  var filter = new Filters();
+  var service = new Api(baseUrl);
```
Cool!  Great refactor :+1:

jniles commented on client/src/modules/patients/patients.service.js:
```diff
@@ -286,5 +300,17 @@ function PatientService($http, util, Session, $uibModal,
   }
 
+  // downloads a type of report based on the
+  service.download = function download(type) {
```
You never actually use this method.  Should it be removed?

lomamech committed:
fix(patient registry)
- Use class grid-full-height-with-filters
- Improve html for searche modal
- change the filter for get report from server
- Fix code and remove unused function
- Use the translate directive

jniles commented on client/src/modules/patients/registry/registry.js:
```diff
@@ -107,9 +107,22 @@ function PatientRegistryController($state, Patients, Notify, AppCache,
   };
 
+  // API register function
+  function onRegisterApi(gridApi) {
```
It doesn't look like you ever use `onRegisterApi()`.  Do you need it?

jniles commented on client/src/modules/patients/registry/registry.js:
```diff
@@ -202,23 +196,35 @@ function PatientRegistryController($state, Patients, Notify, AppCache,
+  }
+
+  // display the patient registry printable report
+  vm.openJournalReport = function openJournalReport() {
```
I don't think this should be called `openJournalReport` since it isn't the journal.  Maybe `PatientReport`?

jniles commented on client/src/modules/patients/registry/search.modal.html:
```diff
@@ -12,134 +12,140 @@
+    <uib-tabset>
+      <uib-tab index="0" heading="{{ 'FORM.LABELS.SEARCH_QUERRIES' | translate}}" data-custom-filter-tab>
+        <div class="form-group" ng-class="{ 'has-error' : ModalForm.$submitted && ModalForm.name.$invalid }">
+          <label class="control-label">
```
Please use the `translate` directive instead of the `translate` filter.

jniles commented on test/end-to-end/patient/registry.search.js:
```diff
@@ -1,21 +1,16 @@
 
-const helpers = require('../shared/helpers');
+const Filters = require('../shared/components/bhFilters');
+const SearchModal = require('../shared/search.page');
```
:+1:  Good usage of Search Modal

jniles reviewed (changes requested):
@lomamech, there are a few changes to be made before this lands in master.

 1. Make sure that you rename code that you copied from other contexts - in the case `openJournal` shouldn't appear in the patient registration controller.
 2. Make sure you use all the methods you define - I made a few comments on what seems to be extra code.
 3. Make sure you use the `translate` directive instead of the filter for performance reasons.

Your use of the search modal page object was excellent.  Great tests :+1:

jniles commented on client/src/modules/patients/patients.service.js:
```diff
@@ -22,23 +22,22 @@ PatientService.$inject = [
 
-  service.read = read;
+  var patientFilters = new Filters();
+  var filterCache = new AppCache('cash-filters');
```
Oh - didn't catch this.  This should be `patient-filters` or something.  If you copied this code directly from the Cash Payments Registry, you will find it using Cash's filters.

sfount replied:
The key here cannot be the same as cash filters - the filters for the cash registry are not all compatible with the patient registry.  This key must change.

jniles commented on client/src/modules/patients/registry/registry.html:
```diff
@@ -7,68 +7,79 @@
     <div
       id="patient-registry"
-      class="grid-full-height"
+      class="grid-full-height-with-filters"
```
Great fix :+1:

jniles reviewed (dismissed):
@lomamech, my other comments look like they have been addressed - well done.

I've caught one more critical bug - the AppCache key needs to change.  It should be unique to the Patients Registry.

After that, code LGTM 👍

sfount commented on client/src/modules/patients/patients.service.js:
```diff
@@ -220,51 +219,66 @@ function PatientService($http, util, Session, $uibModal,
-    });
+  patientFilters.registerDefaultFilters([
+    { key : 'period', label : 'TABLE.COLUMNS.PERIOD', valueFilter : 'translate' },
+    { key : 'custom_period_start', label : 'PERIODS.START', valueFilter : 'date' },
```
It would good to add the comparison here so that the filters display 'Period Start > DD/MM/YYYY' and 'Period End < DD/MM/YYY'. 

You can do this by passing in the `comparitor` attribute, for example: 
```js
{ key : 'custom_period_start', label : 'PERIODS.START', valueFilter : 'date', comparitor : '>'}
```

sfount commented on server/controllers/medical/patients/index.js:
```diff
@@ -373,24 +373,28 @@ function searchByName(req, res, next) {
+  db.convert(options, ['patient_group_uuid', 'debtor_group_uuid', 'debtor_uuid']);
 
-  const filters = new FilterParser(options, { tableAlias : 'p' });
+  const filters = new FilterParser(options, { tableAlias : 'p', autoParseStatements : false  });
```
👍

sfount reviewed (commented):
LGTM. There are 1 or 2 comments that should be addressed. I will pull this and check the functionality!

lomamech committed:
perf(patient Registry) :
- Change the filter key
- Using the comparitor for Period Attribut

lomamech committed:
fix(patient registry)
- Fix error when CSV export
- Applie the filters for all exportations
- Improve patientService

lomamech committed:
Fix (Refactor Patient Registry)
Fix error around the using of filters params display
